Upgrade to future Sleuthkit-4.13 when released

Seems they are implementing BitLocker decryption here:
https://github.com/sleuthkit/sleuthkit/pull/2923

After a lot of build issues of TSK with `mbedtls` (for Bitlocker), the processing worked with a encrypted Bitlocker partition:
```
iped -d Bitlocker.dd -p 661573-082060-422873-216733-054956-025597-647834-485705 -o case
```

If tested with incorrect recovery key, it show this message:
```
[ERROR] [engine.datasource.SleuthkitReader]     2. BitLocker initialization failed (Incorrect password entered (Recovery key identifier: 751213C4-D851-4191-9B58-8D0E69442BEF)) (Sector offset: 32768, Partition Type: Basic data partition) Image: /xxxx/exemplos/bitlocker/Bitlocker.dd
```

Now I'll try to make a solution to compile with mbedtls in a less invasive way

--- a/iped-app/pom.xml
+++ b/iped-app/pom.xml
@@ -353,7 +353,7 @@
                                 <artifactItem>
                                     <groupId>org.sleuthkit</groupId>
                                     <artifactId>sleuthkit-dlls</artifactId>
-                                    <version>4.11.1-p3</version>
+                                    <version>4.15.0-p1</version>
                                     <type>zip</type>
                                     <overWrite>false</overWrite>
                                     <outputDirectory>${tools.dir}</outputDirectory>

--- a/iped-app/resources/config/LocalConfig.txt
+++ b/iped-app/resources/config/LocalConfig.txt
@@ -28,7 +28,7 @@ numThreads = default
 # Full path for sleuthkit-4.x.x.jar built on Linux OS.
 # It is distributed a Windows only version with some optimization patches.
 # It is recommended to apply the patch (in source folder) before building sleuthkit on Linux.
-#tskJarPath = /usr/local/share/java/sleuthkit-4.11.1.jar
+#tskJarPath = /usr/local/share/java/sleuthkit-4.14.0.jar
 
 # Plugin folder for optional libs not embedded because of license restrictions.
 # They will be dynamically loaded at runtime.

--- a/iped-app/resources/config/LocalConfig.txt
+++ b/iped-app/resources/config/LocalConfig.txt
@@ -28,7 +28,7 @@ numThreads = default
 # Full path for sleuthkit-4.x.x.jar built on Linux OS.
 # It is distributed a Windows only version with some optimization patches.
 # It is recommended to apply the patch (in source folder) before building sleuthkit on Linux.
-#tskJarPath = /usr/local/share/java/sleuthkit-4.14.0.jar
+#tskJarPath = /usr/local/share/java/sleuthkit-4.15.0.jar
 
 # Plugin folder for optional libs not embedded because of license restrictions.
 # They will be dynamically loaded at runtime.

--- a/iped-engine/src/main/java/iped/engine/data/IPEDSource.java
+++ b/iped-engine/src/main/java/iped/engine/data/IPEDSource.java
@@ -35,6 +35,7 @@ import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.stream.IntStream;
 
+import org.apache.commons.lang3.StringUtils;
 import org.apache.lucene.analysis.Analyzer;
 import org.apache.lucene.document.Document;
 import org.apache.lucene.index.DirectoryReader;
@@ -49,10 +50,13 @@ import org.apache.lucene.search.DocIdSetIterator;
 import org.apache.lucene.search.IndexSearcher;
 import org.apache.lucene.store.Directory;
 import org.apache.lucene.util.Bits;
+import org.sleuthkit.datamodel.DataSource;
 import org.sleuthkit.datamodel.SleuthkitCase;
 import org.sleuthkit.datamodel.TskCoreException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
 
 import iped.data.IBookmarks;
 import iped.data.IIPEDSource;
@@ -79,6 +83,7 @@ import iped.exception.IPEDException;
 import iped.properties.BasicProps;
 import iped.utils.IOUtil;
 import iped.utils.SelectImagePathWithDialog;
+import iped.utils.UTF8Properties;
 
 public class IPEDSource implements IIPEDSource {
 
@@ -90,6 +95,10 @@ public class IPEDSource implements IIPEDSource {
     public static final String LIB_DIR = "lib"; //$NON-NLS-1$
     public static final String SLEUTH_DB = "sleuth.db"; //$NON-NLS-1$
     public static final String PREV_TEMP_INFO_PATH = DATA_DIR + "/prevTempDir.txt"; //$NON-NLS-1$
+
+    // key in acquisition tool settings; the password for decrypting an image
+    // @see: org.sleuthkit.datamodel.SleuthkitCase.IMAGE_PASSWORD_KEY
+    private static final String IMAGE_PASSWORD_KEY = "imagePassword";
 
     /**
      * workaround para JVM não coletar objeto, nesse caso Sleuthkit perde referencia
@@ -228,7 +237,7 @@ public class IPEDSource implements IIPEDSource {
             populateEvidenceUUIDs();
             countTotalItems();
 
-            SleuthkitReader.loadImagePasswords(moduleDir);
+            migrateImagesPassword();
 
             loadLeafCategories();
             loadCategoryTree();
@@ -253,6 +262,47 @@ public class IPEDSource implements IIPEDSource {
                 throw (RuntimeException) e;
             }
             throw new RuntimeException(e.getMessage(), e);
+        }
+    }
+
+     private void migrateImagesPassword() {
+
+        if (sleuthCase == null) {
+            return;
+        }
+
+        File file = new File(moduleDir, SleuthkitReader.PASSWORD_PER_IMAGE);
+        if (!file.exists()) {
+            return;
+        }
+
+        // This block handles a data migration scenario for backward compatibility.
+        // In previous versions of IPED, passwords for encrypted disk images were stored in a separate properties file.
+        // This code checks if the password is not yet stored in the Sleuthkit database (as part of the DataSource acquisition details).
+        // If it's not, it reads the password from the old properties file and updates the DataSource object,
+        // effectively migrating the password to the new storage mechanism used by TSK (The Sleuth Kit).
+        // This allows IPED to seamlessly open older cases.
+        UTF8Properties props = new UTF8Properties();
+        try {
+            props.load(file);
+            for (DataSource ds : sleuthCase.getDataSources()) {
+                if (!StringUtils.contains(ds.getAcquisitionToolSettings(), IMAGE_PASSWORD_KEY)) {
+                    // if password is not in the DB yet, get it from props and set it
+                    String password = props.getProperty(ds.getName() + "_PASSWORD");
+                    if (password != null) {
+                        Map<String, Object> acquisitionToolMap = new HashMap<>();
+                        acquisitionToolMap.put(IMAGE_PASSWORD_KEY, password);
+                        String acquisitionToolJson = (new Gson()).toJson(acquisitionToolMap);
+
+                        ds.setAcquisitionToolDetails(ds.getAcquisitionToolName(), ds.getAcquisitionToolVersion(),
+                                acquisitionToolJson);
+                    }
+                }
+            }
+        } catch (Exception e) {
+            LOGGER.error(
+                    "Error migrating image passwords from properties file to Sleuthkit database. " +
+                    "This may cause issues opening encrypted images if the password is not already stored in the database: " + casePath, e);
         }
     }
 

--- a/iped-engine/src/main/java/iped/engine/data/IPEDSource.java
+++ b/iped-engine/src/main/java/iped/engine/data/IPEDSource.java
@@ -100,12 +100,6 @@ public class IPEDSource implements IIPEDSource {
     // @see: org.sleuthkit.datamodel.SleuthkitCase.IMAGE_PASSWORD_KEY
     private static final String IMAGE_PASSWORD_KEY = "imagePassword";
 
-    /**
-     * workaround para JVM não coletar objeto, nesse caso Sleuthkit perde referencia
-     * para FS_INFO
-     */
-    private static List<SleuthkitCase> tskCaseList = Collections.synchronizedList(new ArrayList<SleuthkitCase>());
-
     private File casePath;
     private File moduleDir;
     private File index;
@@ -217,8 +211,6 @@ public class IPEDSource implements IIPEDSource {
 
                 if (!isReport)
                     updateImagePathsToAbsolute(casePath, sleuthFile);
-
-                tskCaseList.add(sleuthCase);
             }
 
             AnalysisConfig analysisConfig = ConfigurationManager.get().findObject(AnalysisConfig.class);
@@ -708,7 +700,6 @@ public class IPEDSource implements IIPEDSource {
             // causes "case is closed" error in some cases
             // sleuthCase.close();
             sleuthCase = SleuthkitInputStreamFactory.openSleuthkitCase(tmpCaseFile.getAbsolutePath());
-            tskCaseList.add(sleuthCase);
         }
     }
 

--- a/iped-engine/src/main/java/iped/engine/data/IPEDSource.java
+++ b/iped-engine/src/main/java/iped/engine/data/IPEDSource.java
@@ -21,6 +21,8 @@ package iped.engine.data;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.BitSet;
 import java.util.Collections;
@@ -36,6 +38,7 @@ import java.util.concurrent.Executors;
 import java.util.stream.IntStream;
 
 import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.SystemUtils;
 import org.apache.lucene.analysis.Analyzer;
 import org.apache.lucene.document.Document;
 import org.apache.lucene.index.DirectoryReader;
@@ -55,6 +58,7 @@ import org.sleuthkit.datamodel.SleuthkitCase;
 import org.sleuthkit.datamodel.TskCoreException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.sqlite.SQLiteException;
 
 import com.google.gson.Gson;
 
@@ -196,21 +200,17 @@ public class IPEDSource implements IIPEDSource {
 
             isReport = new File(moduleDir, "data/containsReport.flag").exists(); //$NON-NLS-1$
 
-            File sleuthFile = new File(casePath, SLEUTH_DB);
-            if (sleuthFile.exists()) {
+            Path sleuthFile = casePath.toPath().resolve(SLEUTH_DB);
+            if (Files.exists(sleuthFile)) {
                 if (SleuthkitReader.sleuthCase != null)
                     // workaroud para demora ao abrir o caso enquanto tsk_loaddb não termina
                     sleuthCase = SleuthkitReader.sleuthCase;
                 else {
-                    // Unpatched TSK doesn't work with a read-only DB, must be writeable
-                    if (!SleuthkitReader.isTSKPatched()) {
-                        sleuthFile = SleuthkitInputStreamFactory.getWriteableDBFile(sleuthFile);
-                    }
-                    sleuthCase = SleuthkitInputStreamFactory.openSleuthkitCase(sleuthFile.getAbsolutePath());
+                    sleuthCase = SleuthkitInputStreamFactory.openSleuthkitCase(sleuthFile);
                 }
 
                 if (!isReport)
-                    updateImagePathsToAbsolute(casePath, sleuthFile);
+                    fixImagePathsToLocalEnvironment(casePath, sleuthFile.toFile());
             }
 
             AnalysisConfig analysisConfig = ConfigurationManager.get().findObject(AnalysisConfig.class);
@@ -633,7 +633,7 @@ public class IPEDSource implements IIPEDSource {
         }
     }
 
-    private void updateImagePathsToAbsolute(File casePath, File sleuthFile) throws Exception {
+    private void fixImagePathsToLocalEnvironment(File casePath, File sleuthFile) throws Exception {
         char letter = casePath.getAbsolutePath().charAt(0);
         boolean isWindowsNetworkShare = casePath.getAbsolutePath().startsWith("\\\\");
         Map<Long, List<String>> imgPaths = sleuthCase.getImagePaths();
@@ -647,13 +647,12 @@ public class IPEDSource implements IIPEDSource {
                     if (new File(newPath).exists())
                         newPaths.add(newPath);
                 } else if ((new File(path).exists() && path.contains(File.separator))
-                        || (System.getProperty("os.name").startsWith("Windows")
-                                && path.toLowerCase().contains("physicaldrive"))) {
+                        || (SystemUtils.IS_OS_WINDOWS && path.toLowerCase().contains("physicaldrive"))) {
                     newPaths = null;
                     break;
                 } else {
                     path = path.replace("/", File.separator).replace("\\", File.separator);
-                    String newPath = letter + path.substring(1);
+                    String newPath = new File(path).isAbsolute() && SystemUtils.IS_OS_WINDOWS ? letter + path.substring(1) : path;
                     if (new File(newPath).exists())
                         newPaths.add(newPath);
                     else {
@@ -678,8 +677,7 @@ public class IPEDSource implements IIPEDSource {
             }
             if (newPaths != null)
                 if (newPaths.size() > 0) {
-                    testCanWriteToCase(sleuthFile);
-                    sleuthCase.setImagePaths(id, newPaths);
+                    updateImagePaths(id, newPaths);
                 } else if (iw == null) {
                     if (askImagePathIfNotFound) {
                         askNewImagePath(id, paths, sleuthFile);
@@ -690,16 +688,32 @@ public class IPEDSource implements IIPEDSource {
         }
     }
 
-    File tmpCaseFile = null;
+    private void updateImagePaths(long imgId, List<String> newPaths) throws IOException, TskCoreException {
+        try {
 
-    private void testCanWriteToCase(File sleuthFile) throws TskCoreException, IOException {
-        if (tmpCaseFile != null) return;
-        File writeableDBFile = SleuthkitInputStreamFactory.getWriteableDBFile(sleuthFile);
-        if (writeableDBFile != sleuthFile){
-            tmpCaseFile = writeableDBFile;
-            // causes "case is closed" error in some cases
-            // sleuthCase.close();
-            sleuthCase = SleuthkitInputStreamFactory.openSleuthkitCase(tmpCaseFile.getAbsolutePath());
+            sleuthCase.setImagePaths(imgId, newPaths);
+
+        } catch (TskCoreException e) {
+
+            // This only happens in patched versions of TSK prior to 4.14.0, where the sleuthCase.setImagePaths 
+            // method does not support transient image paths in read-only cases.
+            // The workaround is to create a temporary copy of the Sleuthkit database in a writable location and update the image paths there.
+            if (e.getCause() instanceof SQLiteException && e.getCause().getMessage().contains("SQLITE_READONLY")) {
+
+                // Do not close sleuthCase here. It is cached in SleuthkitInputStreamFactory.sleuthkitCasesMap
+                // and may still be referenced by other code paths.
+                // sleuthCase.close();
+
+                Path currentDBPath = Paths.get(sleuthCase.getDbDirPath(), sleuthCase.getDatabaseName());
+                Path tmpDBPath = SleuthkitInputStreamFactory.getWriteableDBPath(currentDBPath, true);
+                sleuthCase = SleuthkitInputStreamFactory.openSleuthkitCase(tmpDBPath);
+
+                // set newPaths in the new writable sleuthCase instance
+                sleuthCase.setImagePaths(imgId, newPaths);
+
+            } else {
+                throw e;
+            }
         }
     }
 
@@ -718,8 +732,7 @@ public class IPEDSource implements IIPEDSource {
                     throw new IOException(Messages.getString("IPEDSource.ImgFragNotFound") + basePath + ext); //$NON-NLS-1$
                 newPaths.add(basePath + ext);
             }
-        testCanWriteToCase(sleuthFile);
-        sleuthCase.setImagePaths(imgId, newPaths);
+        updateImagePaths(imgId, newPaths);
     }
 
     public String getItemProperty(int id, String propertyName) {

--- a/iped-engine/src/main/java/iped/engine/datasource/SleuthkitReader.java
+++ b/iped-engine/src/main/java/iped/engine/datasource/SleuthkitReader.java
@@ -323,11 +323,6 @@ public class SleuthkitReader extends DataSourceReader {
 
         int sectorSize = args.getBlocksize();
 
-        String password = getEvidencePassword(image);
-        if (password != null) {
-            setImagePassword(output, image.getName(), password);
-        }
-
         firstId = null;
         lastId = null;
         sleuthIdToId.clear();
@@ -422,7 +417,10 @@ public class SleuthkitReader extends DataSourceReader {
                     idRangeMap.notify();
                 }
 
-                addImage = sleuthCase.makeAddImageProcess(timezone, true, false, "");
+                String password = getEvidencePassword(image);
+
+                addImage = sleuthCase.makeAddImageProcess(timezone, true, false, "", password);
+
                 addImageInBackground(addImage, image, sectorSize);
 
             }

--- a/iped-engine/src/main/java/iped/engine/datasource/SleuthkitReader.java
+++ b/iped-engine/src/main/java/iped/engine/datasource/SleuthkitReader.java
@@ -21,6 +21,8 @@ package iped.engine.datasource;
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.sql.Connection;
 import java.sql.DriverManager;
 import java.sql.ResultSet;
@@ -150,8 +152,8 @@ public class SleuthkitReader extends DataSourceReader {
     // via referência interna ao JNI para acessar os itens do caso
     public static volatile SleuthkitCase sleuthCase;
 
-    public static File getSleuthkitDB(File output) {
-        return new File(output.getParent(), DB_NAME);
+    public static Path getSleuthkitDB(File output) {
+        return output.toPath().getParent().resolve(DB_NAME).toAbsolutePath();
     }
 
     public SleuthkitReader(boolean embeddedDisk, ICaseData caseData, File output) {
@@ -264,7 +266,7 @@ public class SleuthkitReader extends DataSourceReader {
         }
     }
 
-    private static synchronized void checkTSKVersion() throws Exception {
+    private static synchronized void checkTSKVersion() {
 
         if (tskChecked) {
             return;
@@ -291,14 +293,14 @@ public class SleuthkitReader extends DataSourceReader {
         int maxMinorVerTested = Integer.valueOf(MAX_TSK_VER_TESTED.split("\\.")[1]);
 
         if (majorVerExpected != majorVerFound || minorVerFound < minorVerExpected)
-            throw new Exception("Sleuthkit version " + tskVer + " not supported. Install version " + MIN_TSK_VER_TESTED); //$NON-NLS-1$ //$NON-NLS-2$
+            throw new RuntimeException("Sleuthkit version " + tskVer + " not supported. Install version " + MIN_TSK_VER_TESTED);
         if (minorVerFound > maxMinorVerTested || UNTESTED_TSK_VERS.contains(tskVer))
             LOGGER.error("Sleuthkit version " + tskVer + " not tested! It may contain incompatibilities!"); //$NON-NLS-1$ //$NON-NLS-2$
 
         tskChecked = true;
     }
     
-    public static synchronized boolean isTSKPatched() throws Exception {
+    public static synchronized boolean isTSKPatched() {
         if (!tskChecked) {
             checkTSKVersion();
         }
@@ -347,7 +349,7 @@ public class SleuthkitReader extends DataSourceReader {
             }
         }
 
-        String dbPath = getSleuthkitDB(output).getAbsolutePath();
+        Path dbPath = getSleuthkitDB(output);
 
         if (listOnly || embeddedDisk) {
 
@@ -356,14 +358,14 @@ public class SleuthkitReader extends DataSourceReader {
             if (sleuthCase == null) {
                 synchronized (this.getClass()) {
                     if (sleuthCase == null) {
-                        if (new File(dbPath).exists()) {
+                        if (Files.exists(dbPath)) {
                             sleuthCase = SleuthkitInputStreamFactory.openSleuthkitCase(dbPath);
 
                         } else {
                             UIPropertyListenerProvider.getInstance().firePropertyChange("mensagem", "", //$NON-NLS-1$ //$NON-NLS-2$
                                     Messages.getString("SleuthkitReader.Creating") + dbPath); //$NON-NLS-1$
                             LOGGER.info("Creating database {}", dbPath); //$NON-NLS-1$
-                            sleuthCase = SleuthkitCase.newCase(dbPath);
+                            sleuthCase = SleuthkitCase.newCase(dbPath.toString());
                             LOGGER.info("{} database created", dbPath); //$NON-NLS-1$
                         }
                     }
@@ -376,7 +378,7 @@ public class SleuthkitReader extends DataSourceReader {
             }
             System.setProperties(sysProps);
 
-            SleuthkitClient.initSleuthkitServers(new File(dbPath));
+            SleuthkitClient.initSleuthkitServers(dbPath.toFile());
 
             Long[] range = getDecodedRangeId(image, output);
             if (range != null && args.isContinue()) {

--- a/iped-engine/src/main/java/iped/engine/datasource/SleuthkitReader.java
+++ b/iped-engine/src/main/java/iped/engine/datasource/SleuthkitReader.java
@@ -27,6 +27,7 @@ import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Statement;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.Date;
 import java.util.HashMap;
@@ -101,7 +102,8 @@ public class SleuthkitReader extends DataSourceReader {
 
     // TODO update @deleteDatasource() when updating TSK
     public static final String MIN_TSK_VER_TESTED = "4.11.0";
-    public static final String MAX_TSK_VER_TESTED = "4.12.0";
+    public static final String MAX_TSK_VER_TESTED = "4.14.0";
+    public static final List<String> UNTESTED_TSK_VERS = Arrays.asList("4.13.0");
 
     public static String DB_NAME = "sleuth.db"; //$NON-NLS-1$
     public static MediaType UNALLOCATED_MIMETYPE = BaseCarveTask.UNALLOCATED_MIMETYPE;
@@ -290,7 +292,7 @@ public class SleuthkitReader extends DataSourceReader {
 
         if (majorVerExpected != majorVerFound || minorVerFound < minorVerExpected)
             throw new Exception("Sleuthkit version " + tskVer + " not supported. Install version " + MIN_TSK_VER_TESTED); //$NON-NLS-1$ //$NON-NLS-2$
-        if (minorVerFound > maxMinorVerTested)
+        if (minorVerFound > maxMinorVerTested || UNTESTED_TSK_VERS.contains(tskVer))
             LOGGER.error("Sleuthkit version " + tskVer + " not tested! It may contain incompatibilities!"); //$NON-NLS-1$ //$NON-NLS-2$
 
         tskChecked = true;

--- a/iped-engine/src/main/java/iped/engine/datasource/SleuthkitReader.java
+++ b/iped-engine/src/main/java/iped/engine/datasource/SleuthkitReader.java
@@ -497,44 +497,6 @@ public class SleuthkitReader extends DataSourceReader {
         }
     }
     
-    public static synchronized void loadImagePasswords(File output) {
-        File file = new File(output, PASSWORD_PER_IMAGE);
-        if (file.exists()) {
-            UTF8Properties props = new UTF8Properties();
-            try {
-                props.load(file);
-                for (String key : props.stringPropertyNames()) {
-                    setEnvVar(key, props.getProperty(key));
-                }
-            } catch (Exception e) {
-                e.printStackTrace();
-            }
-        }
-    }
-
-    private static synchronized void setImagePassword(File output, String imageName, String password) {
-        String envVar = imageName + "_PASSWORD";
-        setEnvVar(envVar, password);
-
-        File file = new File(output, PASSWORD_PER_IMAGE);
-        UTF8Properties props = new UTF8Properties();
-        try {
-            if (file.exists()) {
-                props.load(file);
-            }
-            props.setProperty(envVar, password);
-            props.store(file);
-
-        } catch (IOException e) {
-            e.printStackTrace();
-        }
-    }
-
-    private static void setEnvVar(String envVar, String value) {
-        Util.setEnvVar(envVar, value);
-        SleuthkitClient.addEnvVar(envVar, value);
-    }
-
     public void read(File image) throws Exception {
         read(image, null);
     }

--- a/iped-engine/src/main/java/iped/engine/datasource/SleuthkitReader.java
+++ b/iped-engine/src/main/java/iped/engine/datasource/SleuthkitReader.java
@@ -98,7 +98,7 @@ public class SleuthkitReader extends DataSourceReader {
     private static Logger LOGGER = LoggerFactory.getLogger(SleuthkitReader.class);
 
     private static final String RANGE_ID_FILE = "data/SleuthkitIdsPerImage.txt";
-    private static final String PASSWORD_PER_IMAGE = "data/PasswordPerImage.txt";
+    public static final String PASSWORD_PER_IMAGE = "data/PasswordPerImage.txt";
 
     // TODO update @deleteDatasource() when updating TSK
     public static final String MIN_TSK_VER_TESTED = "4.11.0";

--- a/iped-engine/src/main/java/iped/engine/datasource/SleuthkitReader.java
+++ b/iped-engine/src/main/java/iped/engine/datasource/SleuthkitReader.java
@@ -103,9 +103,9 @@ public class SleuthkitReader extends DataSourceReader {
     public static final String PASSWORD_PER_IMAGE = "data/PasswordPerImage.txt";
 
     // TODO update @deleteDatasource() when updating TSK
-    public static final String MIN_TSK_VER_TESTED = "4.11.0";
-    public static final String MAX_TSK_VER_TESTED = "4.14.0";
-    public static final List<String> UNTESTED_TSK_VERS = Arrays.asList("4.13.0");
+    public static final String MIN_TSK_VER_TESTED = "4.13.0";
+    public static final String MAX_TSK_VER_TESTED = "4.15.0";
+    public static final List<String> UNTESTED_TSK_VERS = Arrays.asList("4.13.", "4.14.");
 
     public static String DB_NAME = "sleuth.db"; //$NON-NLS-1$
     public static MediaType UNALLOCATED_MIMETYPE = BaseCarveTask.UNALLOCATED_MIMETYPE;
@@ -293,8 +293,8 @@ public class SleuthkitReader extends DataSourceReader {
         int maxMinorVerTested = Integer.valueOf(MAX_TSK_VER_TESTED.split("\\.")[1]);
 
         if (majorVerExpected != majorVerFound || minorVerFound < minorVerExpected)
-            throw new RuntimeException("Sleuthkit version " + tskVer + " not supported. Install version " + MIN_TSK_VER_TESTED);
-        if (minorVerFound > maxMinorVerTested || UNTESTED_TSK_VERS.contains(tskVer))
+            throw new RuntimeException("Sleuthkit version " + tskVer + " not supported. Install version " + MAX_TSK_VER_TESTED);
+        if (minorVerFound > maxMinorVerTested || UNTESTED_TSK_VERS.stream().anyMatch(tskVer::startsWith))
             LOGGER.error("Sleuthkit version " + tskVer + " not tested! It may contain incompatibilities!"); //$NON-NLS-1$ //$NON-NLS-2$
 
         tskChecked = true;

--- a/iped-engine/src/main/java/iped/engine/sleuthkit/SleuthkitClient.java
+++ b/iped-engine/src/main/java/iped/engine/sleuthkit/SleuthkitClient.java
@@ -125,10 +125,6 @@ public class SleuthkitClient implements Comparable<SleuthkitClient> {
             requestTime = 0;
     }
 
-    public static synchronized void addEnvVar(String key, String value) {
-        newEnvVars.put(key, value);
-    }
-
     public static SleuthkitClient get() {
 
         synchronized (lock) {

--- a/iped-engine/src/main/java/iped/engine/sleuthkit/SleuthkitInputStreamFactory.java
+++ b/iped-engine/src/main/java/iped/engine/sleuthkit/SleuthkitInputStreamFactory.java
@@ -17,6 +17,7 @@ import org.sqlite.SQLiteException;
 
 import iped.engine.config.ConfigurationManager;
 import iped.engine.config.FileSystemConfig;
+import iped.engine.datasource.SleuthkitReader;
 import iped.io.SeekableInputStream;
 import iped.utils.EmptyInputStream;
 import iped.utils.IOUtil;
@@ -69,12 +70,9 @@ public class SleuthkitInputStreamFactory extends SeekableInputStreamFactory {
                 if (sleuthkitCase == null) {
                     try {
                         File tskDB = Paths.get(dataSource).toFile();
-                        // Workaround for https://github.com/sepinf-inc/IPED/issues/2799
-                        // Condition below was commented for now, but probably there is a better
-                        // solution than always copying the DB, which should be avoided when possible
-                        //if (!SleuthkitReader.isTSKPatched()) {
-                        tskDB = getWriteableDBFile(tskDB);
-                        //}
+                        if (!SleuthkitReader.isTSKPatched()) {
+                            tskDB = getWriteableDBFile(tskDB);
+                        }
                         sleuthkitCase = openSleuthkitCase(tskDB.getAbsolutePath());
 
                     } catch (Exception e) {
@@ -91,7 +89,6 @@ public class SleuthkitInputStreamFactory extends SeekableInputStreamFactory {
      * 
      * @param tskDB
      * @return
-     * @throws TskCoreException
      */
     public static SleuthkitCase openSleuthkitCase(String tskDBPath) {
         return sleuthkitCaseMap.computeIfAbsent(tskDBPath, (key) -> {

--- a/iped-engine/src/main/java/iped/engine/sleuthkit/SleuthkitInputStreamFactory.java
+++ b/iped-engine/src/main/java/iped/engine/sleuthkit/SleuthkitInputStreamFactory.java
@@ -2,11 +2,14 @@ package iped.engine.sleuthkit;
 
 import java.io.File;
 import java.io.IOException;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
 import java.util.Map.Entry;
 import java.util.Properties;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
 
 import org.sleuthkit.datamodel.Content;
 import org.sleuthkit.datamodel.SleuthkitCase;
@@ -27,7 +30,7 @@ public class SleuthkitInputStreamFactory extends SeekableInputStreamFactory {
 
     private static final Logger logger = LoggerFactory.getLogger(SleuthkitInputStreamFactory.class);
 
-    private static final ConcurrentHashMap<String, SleuthkitCase> sleuthkitCaseMap = new ConcurrentHashMap<>();
+    private static final ConcurrentHashMap<Path, SleuthkitCase> sleuthkitCaseMap = new ConcurrentHashMap<>();
 
     private SleuthkitCase sleuthkitCase;
     private Content content;
@@ -69,11 +72,8 @@ public class SleuthkitInputStreamFactory extends SeekableInputStreamFactory {
             synchronized (this) {
                 if (sleuthkitCase == null) {
                     try {
-                        File tskDB = Paths.get(dataSource).toFile();
-                        if (!SleuthkitReader.isTSKPatched()) {
-                            tskDB = getWriteableDBFile(tskDB);
-                        }
-                        sleuthkitCase = openSleuthkitCase(tskDB.getAbsolutePath());
+                        logger.info("Opening Sleuthkit case for data source: " + dataSource);
+                        sleuthkitCase = openSleuthkitCase(Paths.get(dataSource));
 
                     } catch (Exception e) {
                         throw new RuntimeException(e);
@@ -90,30 +90,47 @@ public class SleuthkitInputStreamFactory extends SeekableInputStreamFactory {
      * @param tskDB
      * @return
      */
-    public static SleuthkitCase openSleuthkitCase(String tskDBPath) {
-        return sleuthkitCaseMap.computeIfAbsent(tskDBPath, (key) -> {
+    public synchronized static SleuthkitCase openSleuthkitCase(Path tskDBPath) throws IOException, TskCoreException {
+        Path normalizedDBPath = tskDBPath.toAbsolutePath().normalize();
+
+        if (!SleuthkitReader.isTSKPatched()) {
+            normalizedDBPath = getWriteableDBPath(normalizedDBPath, false).toAbsolutePath().normalize();
+        }
+
+        return sleuthkitCaseMap.computeIfAbsent(normalizedDBPath, (key) -> {
+
             Properties sysProps = System.getProperties();
             try {
-                SleuthkitCase sleuthkitCase = SleuthkitCase.openCase(tskDBPath);
+                SleuthkitCase sleuthkitCase = SleuthkitCase.openCase(key.toString());
+
+                return sleuthkitCase;
+
+            } catch (TskCoreException e) {
+                throw new RuntimeException(e);
+
+            } finally {
+
+                // restore system properties after openCase
                 for (Entry<Object, Object> entry : System.getProperties().entrySet()) {
                     sysProps.setProperty(entry.getKey().toString(), entry.getValue().toString());
                 }
                 System.setProperties(sysProps);
-                return sleuthkitCase;
-            } catch (TskCoreException e) {
-                throw new RuntimeException(e);
             }
         });
     }
 
-    public static File getWriteableDBFile(File sleuthkitDB) throws IOException {
-        if (!IOUtil.canWrite(sleuthkitDB) || !IOUtil.canCreateFile(sleuthkitDB.getParentFile())) {
-            File tmpCaseFile = new File(System.getProperty("java.io.basetmpdir"), //$NON-NLS-1$
-                    "sleuthkit-" + sleuthkitDB.lastModified() + ".db"); //$NON-NLS-1$
-            if (!tmpCaseFile.exists() || tmpCaseFile.length() != sleuthkitDB.length()) {
-                IOUtil.copyFile(sleuthkitDB, tmpCaseFile);
+    public static Path getWriteableDBPath(Path sleuthkitDB, boolean alwaysGetCopy) throws IOException {
+        if (alwaysGetCopy || !IOUtil.canWrite(sleuthkitDB.toFile()) || !IOUtil.canCreateFile(sleuthkitDB.getParent().toFile())) {
+
+            String tmpCaseDBName = "sleuthkit-" + Files.getLastModifiedTime(sleuthkitDB).to(TimeUnit.SECONDS) + ".db";
+            Path tmpCasePath = Paths.get(System.getProperty("java.io.basetmpdir"), tmpCaseDBName);
+
+            if (!Files.exists(tmpCasePath) || Files.size(tmpCasePath) != Files.size(sleuthkitDB)) {
+                Files.copy(sleuthkitDB, tmpCasePath, StandardCopyOption.REPLACE_EXISTING);
             }
-            return tmpCaseFile;
+            tmpCasePath.toFile().setWritable(true);
+
+            return tmpCasePath;
         }
         return sleuthkitDB;
     }

--- a/iped-engine/src/main/java/iped/engine/sleuthkit/SleuthkitInputStreamFactory.java
+++ b/iped-engine/src/main/java/iped/engine/sleuthkit/SleuthkitInputStreamFactory.java
@@ -6,6 +6,7 @@ import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.Map.Entry;
 import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
 
 import org.sleuthkit.datamodel.Content;
 import org.sleuthkit.datamodel.SleuthkitCase;
@@ -24,6 +25,8 @@ import iped.utils.SeekableInputStreamFactory;
 public class SleuthkitInputStreamFactory extends SeekableInputStreamFactory {
 
     private static final Logger logger = LoggerFactory.getLogger(SleuthkitInputStreamFactory.class);
+
+    private static final ConcurrentHashMap<String, SleuthkitCase> sleuthkitCaseMap = new ConcurrentHashMap<>();
 
     private SleuthkitCase sleuthkitCase;
     private Content content;
@@ -90,14 +93,20 @@ public class SleuthkitInputStreamFactory extends SeekableInputStreamFactory {
      * @return
      * @throws TskCoreException
      */
-    public static SleuthkitCase openSleuthkitCase(String tskDBPath) throws TskCoreException {
-        Properties sysProps = System.getProperties();
-        SleuthkitCase sleuthkitCase = SleuthkitCase.openCase(tskDBPath);
-        for (Entry<Object, Object> entry : System.getProperties().entrySet()) {
-            sysProps.setProperty(entry.getKey().toString(), entry.getValue().toString());
-        }
-        System.setProperties(sysProps);
-        return sleuthkitCase;
+    public static SleuthkitCase openSleuthkitCase(String tskDBPath) {
+        return sleuthkitCaseMap.computeIfAbsent(tskDBPath, (key) -> {
+            Properties sysProps = System.getProperties();
+            try {
+                SleuthkitCase sleuthkitCase = SleuthkitCase.openCase(tskDBPath);
+                for (Entry<Object, Object> entry : System.getProperties().entrySet()) {
+                    sysProps.setProperty(entry.getKey().toString(), entry.getValue().toString());
+                }
+                System.setProperties(sysProps);
+                return sleuthkitCase;
+            } catch (TskCoreException e) {
+                throw new RuntimeException(e);
+            }
+        });
     }
 
     public static File getWriteableDBFile(File sleuthkitDB) throws IOException {

--- a/iped-engine/src/main/java/iped/engine/sleuthkit/SleuthkitServer.java
+++ b/iped-engine/src/main/java/iped/engine/sleuthkit/SleuthkitServer.java
@@ -9,6 +9,7 @@ import java.io.RandomAccessFile;
 import java.nio.MappedByteBuffer;
 import java.nio.channels.FileChannel;
 import java.nio.channels.FileChannel.MapMode;
+import java.nio.file.Paths;
 import java.util.HashMap;
 import java.util.HashSet;
 
@@ -76,7 +77,7 @@ public class SleuthkitServer {
             cm.loadConfigs();
             Configuration.getInstance().loadNativeLibs();
 
-            SleuthkitCase sleuthCase = SleuthkitInputStreamFactory.openSleuthkitCase(dbPath);
+            SleuthkitCase sleuthCase = SleuthkitInputStreamFactory.openSleuthkitCase(Paths.get(dbPath));
             HashMap<Long, SleuthkitInputStream> sisMap = new HashMap<>();
 
             java.util.logging.Logger.getLogger("org.sleuthkit").setLevel(java.util.logging.Level.SEVERE); //$NON-NLS-1$

--- a/pom.xml
+++ b/pom.xml
@@ -34,7 +34,7 @@
         <tika.core.version>2.4.0-p1</tika.core.version>
         <pdfbox.version>2.0.27</pdfbox.version>
         <lucene.version>9.2.0</lucene.version>
-        <sleuthkit.version>4.12.0.p1</sleuthkit.version>
+        <sleuthkit.version>4.14.0.p1</sleuthkit.version>
         <libreoffice.version>7.2.2</libreoffice.version>
         <slf4j.version>1.7.25</slf4j.version>
         <log4j.version>2.17.1</log4j.version>

--- a/pom.xml
+++ b/pom.xml
@@ -42,7 +42,7 @@
         <hamcrest.version>3.0</hamcrest.version>
         <icepdf.version>7.0.0</icepdf.version>
         <java.dbx.version>1.1-p6</java.dbx.version>
-        <xerial.sqlite.version>3.41.2.2</xerial.sqlite.version>
+        <xerial.sqlite.version>3.51.3.0</xerial.sqlite.version>
         <dockingframes.version>1.1.2</dockingframes.version>
         <telegram.version>1.0.14</telegram.version>
         <twelvemonkeys.version>3.10.1</twelvemonkeys.version>

--- a/pom.xml
+++ b/pom.xml
@@ -34,7 +34,7 @@
         <tika.core.version>2.4.0-p1</tika.core.version>
         <pdfbox.version>2.0.27</pdfbox.version>
         <lucene.version>9.2.0</lucene.version>
-        <sleuthkit.version>4.14.0.p1</sleuthkit.version>
+        <sleuthkit.version>4.15.0.p1</sleuthkit.version>
         <libreoffice.version>7.2.2</libreoffice.version>
         <slf4j.version>1.7.25</slf4j.version>
         <log4j.version>2.17.1</log4j.version>

